projects/wizard_of_wikipedia/generator/agents.py

```diff
@@ -0,0 +1,358 @@
+        self.knowledge_alpha = opt['knowledge_alpha']
+
+    def _dummy_batch(self, bsz, maxlen):
+        batch = super()._dummy_batch(bsz, maxlen)
+        batch['know_vec'] = th.zeros(bsz, 2, 2).long().cuda()
+        batch['ck_mask'] = th.ones(bsz, 2).byte().cuda()
+        batch['cs_ids'] = th.zeros(bsz).long().cuda()
+        batch['use_cs_ids'] = True
+        return batch
+
+    def compute_loss(self, batch, return_output=False):
+        # first compute our regular forced decoding loss
+        token_loss, model_output = super().compute_loss(batch, return_output=True)
+
+        encoder_states = model_output[2]
+        ctx_know_attn = encoder_states[2]
+
+        if self.knowledge_alpha == 0.0:
+            loss = token_loss
+        else:
+            _, know_pred = ctx_know_attn.max(1)
+            know_acc = (know_pred == batch.cs_ids).float().sum().item()
+            know_chance = batch.ck_mask.sum(1).float().reciprocal().sum().item()
+            self.metrics['know_chance'] += know_chance
+            self.metrics['bsz'] += batch.text_vec.size(0)
+            self.metrics['know_acc'] += know_acc
+            know_loss = th.nn.functional.cross_entropy(
+                ctx_know_attn,
+                batch.cs_ids,
+                reduction='mean',
+            )
+            self.metrics['know_loss'] += know_loss.item() * batch.text_vec.size(0)
+            loss = (
+                (1 - self.knowledge_alpha) * token_loss +
+                self.knowledge_alpha * know_loss
+            )
+        if return_output:
+            return (loss, model_output)
+        else:
+            return loss
+
+    def reset_metrics(self):
+        super().reset_metrics()
+        self.metrics['bsz'] = 0.0
+        self.metrics['know_acc'] = 0.0
+        self.metrics['know_loss'] = 0.0
+        self.metrics['know_chance'] = 0.0
+
+    def report(self):
+        r = super().report()
+        bsz = max(self.metrics['bsz'], 1)
+        for k in ['know_loss', 'know_acc', 'know_chance']:
+            # round and average across all items since last report
+            self.metrics[k] = round_sigfigs(self.metrics[k] / bsz, 4)
+        return r
+
+    def _parse_knowledge(self, obs):
+        if 'knowledge_parsed' in obs:
+            # make a copy of the list to prevent the future padding step from
+            # being destructive
+            return list(obs['knowledge_parsed'])
+
+        checked_sentence = '{} {} {}'.format(
+            obs['title'], TOKEN_KNOWLEDGE, obs['checked_sentence']
+        )
+        # grab all the nonempty knowledge
+        obs_know = [k.strip() for k in obs.get('knowledge', '').split('\n')]
+        obs_know = [k for k in obs_know if k]
+
+        # we want the correct knowledge to always be in index 0
+        try:
+            i = obs_know.index(checked_sentence)
+        except ValueError:
+            # uh oh, couldn't find the sentence in the knowledge. This happens for
```

Comment:
does this really happen still??

Reply:
Yes, I think it happens once. That particular sentence didn't have a title associated or something.